### PR TITLE
Closes #2434 - Strings Hashing bugfix

### LINKED ISSUE #2434: hashing columns containing strings
**Describe the bug**
```python
a = ak.array(['a','b'])
b = ak.array([1,2])

ak.hash([a,b])  # fails

ak.hash(a)  # this also fails, but isn't a problem because a.hash() still works
```

**Error Message**
from typeguard
```
TypeError: type of argument "pda" must be one of (pdarray, List); got list instead
```

**Is this a Blocking Issue**
workaround available for temporary work.

**Additional context**
seg_array also issue

### PR DESCRIPTION
**Pull request by joshmarshall1**
This PR closes #2434 

Patches a bug where Strings being passed into the hash api would fail. Also adds support for Categorical and SegArrays being hashed using their codes and values respectively. 

Due to mypy and circular imports in this file, I added the `hashable` type that is comprised of a `Union` of `pdarray, Strings, Categorical, and Segarray`. This was mainly due to the `# type: ignore` decorator causing the argument in the function definition to extend beyond the 105 character limit.

### CODE REVIEW

**Review by Ethan-DeBandi99 — CHANGES_REQUESTED** [review]
It does not appear that there is any testing for hashing `Strings`. That needs to be added before this is merged (especially because it is a bugfix).

File: arkouda/numeric.py
Comment (pierce314159) [inline arkouda/numeric.py:488]: also i know i signed off hashing the values for segarray during initial design decisions, but after playing around a bit more, I don't think this is what we want to do. I think ultimately segarrays are gonna be difficult to handle (for reasons im about to get into). I think we're gonna wanna do some brainstorming for them, but they are def outside the scope of this bug fix and we should put in a separate issue to handle them

```python
# consider this segarray
>>> sa = ak.segarray(ak.array([0,3,6]), ak.array([0,1,2,3,4,5,0,1,2]))
>>> sa
SegArray([
[0 1 2]
[3 4 5]
[0 1 2]
])

# intuitively the first and third elem should has to the same thing
# since they are the same array and the second should be it's own thing

# we can't get this behavior by hashing the segments since 
# they don't capture the fact that sa[0] and sa[2] are identical
>>> sa.segments
array([0 3 6])

>>> ak.hash(sa.segments)
(array([14020738532124324502 10334800848021183012 12294710806803928155]),
 array([13660161615634858013 14420950124940837019 318151776295263894]))

# we can't hash the values cause they don't really match back up to the original segarray
>>> ak.hash(sa.values)
(array([14020738532124324502 17469661437632883388 2223077377302025035 10334800848021183012 12899154897215201871 1827545583036225533 14020738532124324502 17469661437632883388 2223077377302025035]),
 array([13660161615634858013 100355499037336788 7178333632624491744 14420950124940837019 4776206602043303594 14424044815130752497 13660161615634858013 100355499037336788 7178333632624491744]))

# we thought about hashing each sa[i] and then hashing the resulting pdarray
# but ak.hash(pdarray) returns a pdarray not a single value

# we can't just concatenate the values within a segment because how do u differentiate between 
# [1, 11, 2] and [11, 1, 2] for instance

# one absolutely awful idea that technically works... is to interpret each inner array as a string and then hash that
# (shown here using __repr__ but we would have to do cast server side because repr truncates)
>>> ak.hash(ak.array([ak.array(sa[i]).__repr__() for i in range(3)]))
(array([16660294113708371871 6932019251617944032 16660294113708371871]),
 array([10308859934083080982 18056049045803087039 10308859934083080982]))
# we see that the first and last do match, but it's such an ugly solution
# If we go this route maybe we do hex representation to shorten string lengths
```
These are just some quick ideas, but doing this well is gonna require some thought. I vote we hold off on segarray hashing for now

**Review by pierce314159 — CHANGES_REQUESTED** [review]
I agree we should add some testing for this functionality. I suggested a helper func that will reduce the number of lines needed. And found that segarray hashing is gonna be complicated and we might wanna hold off on that for now

File: arkouda/numeric.py
Comment (pierce314159) [reply arkouda/numeric.py:488]: This has been captured in #2459

File: arkouda/numeric.py
Comment (joshmarshall1) [reply arkouda/numeric.py]: Added the helper function without the SegArray portion. That will be added in #2459

File: arkouda/numeric.py
Comment (joshmarshall1) [reply arkouda/numeric.py]: Removed SegArray handling from this PR

**Comment by joshmarshall1** [comment]
I modified the handling of `Categoricals` after realizing two `Categoricals` with completely different values could have identical `codes` arrays. It was decided the best way to handle this, at least for now, is to use `categorical.categories[categorical.codes]` to generate a `Strings` object that will be unique for every variation of Categoricals.

File: tests/numeric_test.py
Comment (pierce314159) [inline tests/numeric_test.py:240]:
```suggestion
        h1 = ak.hash(ak.Categorical(ak.array(["a", "a", "b", "b", "c"])))
        h2 = ak.hash(ak.Categorical(ak.array(["d", "d", "e", "e", "f"])))
```

**Review by pierce314159 — APPROVED** [review]
Looks good to me

File: arkouda/numeric.py
Comment (pierce314159) [inline arkouda/numeric.py:503]: Why did we got with the command `efuncArr` here? I feel like `"hashlist"` or something like that would be more informative

**Comment by pierce314159** [comment]
> I modified the handling of `Categoricals` after realizing two `Categoricals` with completely different values could have identical `codes` arrays. It was decided the best way to handle this, at least for now, is to use `categorical.categories[categorical.codes]` to generate a `Strings` object that will be unique for every variation of Categoricals.

I did some thinking on this, and it's way more efficient to hash the categories first then expand rather than to expand first and then hash

```python
>>> def test(j):
   ...:      import pandas as pd
   ...:      categories = ak.array([f'str {i}' for i in range(3)])
   ...:      codes = ak.randint(0, 3, 10**j)
   ...:
   ...:      timer1 = pd.Timestamp.now()
   ...:      expand_first_hi, expand_first_lo = ak.hash(categories[codes])
   ...:      timer2 = pd.Timestamp.now()
   ...:      print('expand then hash time')
   ...:      print(timer2-timer1)
   ...:
   ...:      timer1 = pd.Timestamp.now()
   ...:      h = ak.hash(categories)
   ...:      hash_first_hi, hash_first_lo = h[0][codes], h[1][codes]
   ...:      timer2 = pd.Timestamp.now()
   ...:      print('hash then expand time')
   ...:      print(timer2-timer1)
   ...:
   ...:      print()
   ...:      print(f'hash high matches: {ak.all(expand_first_hi == hash_first_hi)}')
   ...:      print(f'hash low matches: {ak.all(expand_first_lo == hash_first_lo)}')
   ...:

>>> test(8)
expand then hash time
0 days 00:00:10.223829
hash then expand time
0 days 00:00:01.357992

hash high matches: True
hash low matches: True
```
These are `10**8` long with comm=none

File: arkouda/numeric.py
Comment (pierce314159) [inline arkouda/numeric.py:432]: we should update how we handle cats to hash the categories and then do integer indexing instead

**Review by pierce314159 — CHANGES_REQUESTED** [review]

File: arkouda/numeric.py
Comment (pierce314159) [reply arkouda/numeric.py:432]: I think we should handle adding categoricals in a separate issue too, since we really want to do this idexing and hashing in the chapel code (so it can be easily called by `hashArrays`

**Comment by pierce314159** [comment]
closing in favor of PR #2475 and issue #2462

**Comment by pierce314159** [comment]
closing in favor of PR https://github.com/Bears-R-Us/arkouda/pull/2475 and issue https://github.com/Bears-R-Us/arkouda/issues/2462

**Pull request closed without merging**
